### PR TITLE
ruby - generate map

### PR DESCRIPTION
**Pull request by Shivansh-yadav13**
-Closes https://github.com/appwrite/appwrite/issues/4134

### CODE REVIEW

**Comment by Shivansh-yadav13** [comment]
@Meldiron please add review

**Comment by PineappleIOnic** [comment]
Thank you so much for the PR 🤩. We're adding the `hacktoberfest-accepted` label to ensure this PR counts towards your Hacktoberfest contributions count. With that said, please stay active on this PR to address any comments once you receive a review. Happy Hacktoberfest! 🎃

File: ruby/generate-map/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+```json
+{
+    "success":true,
+    "image":"iVBORw0KGgoAAAANSUhEUgAAAaQAAALiCAY...QoH9hbkTPQAAAABJRU5ErkJggg=="
+}
+```
+
+## 📝 Environment Variables
+List of environment variables used by this cloud function:
```
Comment (stnguyen90) [inline ruby/generate-map/README.md:26]: Let's remove references to Environment:

```suggestion
## 📝 Variables
List of variables used by this cloud function:
```

File: ruby/generate-map/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+2. Enter this function folder and build the code:
+```
+docker run --rm --interactive --tty --volume $PWD:/usr/code openruntimes/ruby:3.1 sh /usr/local/src/build.sh
+```
+As a result, a `code.tar.gz` file will be generated.
+
+3. Start the Open Runtime:
+```
+docker run -p 3000:3000 -e INTERNAL_RUNTIME_KEY=secret-key -e INTERNAL_RUNTIME_ENTRYPOINT=index.rb --rm --interactive --tty --volume $PWD/code.tar.gz:/tmp/code.tar.gz:ro openruntimes/ruby:3.1 sh /usr/local/src/start.sh
```
Comment (stnguyen90) [inline ruby/generate-map/README.md:50]: Please make sure to use the latest runtime image:

````suggestion
docker run --rm --interactive --tty --volume $PWD:/usr/code openruntimes/ruby:v2-3.1 sh /usr/local/src/build.sh
```
As a result, a `code.tar.gz` file will be generated.

3. Start the Open Runtime:
```
docker run -p 3000:3000 -e INTERNAL_RUNTIME_KEY=secret-key -e INTERNAL_RUNTIME_ENTRYPOINT=index.rb --rm --interactive --tty --volume $PWD/code.tar.gz:/tmp/code.tar.gz:ro openruntimes/ruby:v2-3.1 sh /usr/local/src/start.sh
````

File: ruby/generate-map/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+Your function is now listening on port `3000`, and you can execute it by sending `POST` request with appropriate authorization headers. To learn more about runtime, you can visit Ruby runtime [README](https://github.com/open-runtimes/open-runtimes/tree/main/runtimes/ruby-3.1).
+
+## 📝 Notes
+ - This function is designed for use with Appwrite Cloud Functions. You can learn more about it in [Appwrite docs](https://appwrite.io/docs/functions).
+ - This example is compatible with Ruby 3.1. Other versions may work but are not guaranteed to work as they haven't been tested.
```
Comment (stnguyen90) [inline ruby/generate-map/README.md:57]: No need for the leading spaces

```suggestion
- This function is designed for use with Appwrite Cloud Functions. You can learn more about it in [Appwrite docs](https://appwrite.io/docs/functions).
- This example is compatible with Ruby 3.1. Other versions may work but are not guaranteed to work as they haven't been tested.
```

File: ruby/generate-map/index.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
+require 'open-street-map'
+require 'json'
+require 'uri'
+require 'net/http'
+require 'base64'
+
+def error(res, message)
+  return res.json({"success": false, "message": message})
+end
```
Comment (stnguyen90) [inline ruby/generate-map/index.rb:9]: nice approach to reducing duplicate code!

File: ruby/generate-map/index.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    return error(res, "The longitude must be between -180 and 180")
+  elsif (!(-90..90).include? lat)
+    return error(res, "The latitude must be between -90 and 90")
+  end
+  # Check if User-Agent env var exists
+  user_agent_header = nil
+  begin
+    user_agent_header = req.env['USER_AGENT']
+  rescue
+    return error(res, "Missing 'User-Agent' enviornment variable.")
```
Comment (stnguyen90) [inline ruby/generate-map/index.rb:44]: Let's make this match the variable:

```suggestion
    return error(res, "Missing 'USER_AGENT' function variable.")
```

File: ruby/generate-map/index.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
+require 'open-street-map'
```
Comment (stnguyen90) [inline ruby/generate-map/index.rb:1]: I seem to get an error running this code:

```shell
$ curl -H "X-Internal-Challenge: secret-key" \
>     -H "Content-Type: application/json" \
>     -X POST http://localhost:3000/ \
>     -d '{"payload": "{}"}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4723  100  4706  100    17   135k    500 --:--:-- --:--:-- --:--:--  135k
{
  "code": 500,
  "message": "/usr/code-start/index.rb:1:in `require'\n/usr/code-start/index.rb:1:in `<top (required)>'\n/usr/local/src/server.rb:74:in `load'\n/usr/local/src/server.rb:74:in `block in <top (required)>'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1684:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1684:in `block in compile!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1002:in `block (3 levels) in route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1020:in `route_eval'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1002:in `block (2 levels) in route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1051:in `block in process_route'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1049:in `catch'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1049:in `process_route'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1000:in `block in route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:997:in `each'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:997:in `route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1121:in `block in dispatch!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `catch'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `invoke'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1116:in `dispatch!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:932:in `block in call!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `catch'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `invoke'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:932:in `call!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:921:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/xss_header.rb:20:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/path_traversal.rb:18:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/json_csrf.rb:28:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/base.rb:53:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/base.rb:53:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/frame_options.rb:33:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/logger.rb:17:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/common_logger.rb:38:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:261:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:254:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/head.rb:12:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/method_override.rb:24:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:219:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1995:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1554:in `block in call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1770:in `synchronize'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1554:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/configuration.rb:268:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:89:in `block in handle_request'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:343:in `with_force_shutdown'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:88:in `handle_request'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:430:in `process_client'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:233:in `block in run'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:150:in `block in spawn_thread'"
}
```

in the runtime logs, i see:

```
#<LoadError: cannot load such file -- open-street-map>
```

Are you having this problem?

File: ruby/generate-map/index.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    return error(res, "The longitude and latitude must be a number")
+  elsif (!(-180..180).include? lng)
+    return error(res, "The longitude must be between -180 and 180")
+  elsif (!(-90..90).include? lat)
+    return error(res, "The latitude must be between -90 and 90")
+  end
+  # Check if User-Agent env var exists
+  user_agent_header = nil
+  begin
+    user_agent_header = req.env['USER_AGENT']
```
Comment (stnguyen90) [inline ruby/generate-map/index.rb:42]: Please make sure you're accessing variables via `req.variables`

**Review by stnguyen90 — CHANGES_REQUESTED** [review]
Great PR! 🤯 We left some comments during the review, please check them out.

**Comment by stnguyen90** [comment]
@Shivansh-yadav13 have you had a chance to address my comments? Also, it looks like there's a merge conflict now.

**Comment by Meldiron** [comment]
Hey there 👋
We are temporarily closing your PR as there has been no activity on it. Feel free to re-open it when you're ready to start working on it again 😇

**Pull request closed without merging**